Content.Server/Radiation/Systems/RadiationSystem.cs

```diff
@@ -32,8 +37,91 @@ public override void Initialize()
 
         _blockerQuery = GetEntityQuery<RadiationBlockingContainerComponent>();
         _resistanceQuery = GetEntityQuery<RadiationGridResistanceComponent>();
-        _gridQuery = GetEntityQuery<MapGridComponent>();
         _stackQuery = GetEntityQuery<StackComponent>();
+
+        SubscribeLocalEvent<RadiationSourceComponent, ComponentInit>(OnSourceInit);
+        SubscribeLocalEvent<RadiationSourceComponent, ComponentShutdown>(OnSourceShutdown);
+        SubscribeLocalEvent<RadiationSourceComponent, MoveEvent>(OnSourceMove);
+        SubscribeLocalEvent<RadiationSourceComponent, StackCountChangedEvent>(OnSourceStackChanged);
+        
+        SubscribeLocalEvent<RadiationReceiverComponent, ComponentInit>(OnReceiverInit);
+        SubscribeLocalEvent<RadiationReceiverComponent, ComponentShutdown>(OnReceiverShutdown);
+    }
+
+    private void OnSourceInit(EntityUid uid, RadiationSourceComponent component, ComponentInit args)
+    {
+        UpdateSource(uid, component);
+    }
+
+    private void OnSourceShutdown(EntityUid uid, RadiationSourceComponent component, ComponentShutdown args)
+    {
+        if (_sourceDataMap.Remove(uid))
+        {
+            _sourceTree.Remove(uid);
+        }
+    }
+
+    private void OnSourceMove(EntityUid uid, RadiationSourceComponent component, ref MoveEvent args)
+    {
+        if (args.NewPosition.EntityId == args.OldPosition.EntityId &&
+            args.NewPosition.Position.EqualsApprox(args.OldPosition.Position))
+            return;
+
+        UpdateSource(uid, component, args.Component);
+    }
+
+    private void OnSourceStackChanged(EntityUid uid, RadiationSourceComponent component, StackCountChangedEvent args)
+    {
+        UpdateSource(uid, component);
+    }
+
+    private void OnReceiverInit(EntityUid uid, RadiationReceiverComponent component, ComponentInit args)
+    {
+        _activeReceivers.Add(uid);
+    }
+
+    private void OnReceiverShutdown(EntityUid uid, RadiationReceiverComponent component, ComponentShutdown args)
+    {
+        _activeReceivers.Remove(uid);
+    }
+
+    private void UpdateSource(EntityUid uid, RadiationSourceComponent component, TransformComponent? xform = null)
+    {
+        if (!Resolve(uid, ref xform))
+            return;
```

Comment:
And then from there you can do:
```suggestion
    private void UpdateSource(Entity<RadiationSourceComponent> entity, TransformComponent? xform = null)
    {
        if (!Resolve(entity.Owner, ref xform))
            return;
```

Reply:
All entities have a TransformComponent, so the only case where Transform(uid) fails is when the entity is deleted or doesn't exist.